---
title: "MS3"
author: "Clara Cardoso"
date: "28.08.2020"
output: html_notebook
---

```{r load packages}
Sys.time()
Sys.Date()
getwd()

library(dada2); packageVersion("dada2")
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")
library(decontam); packageVersion("decontam")
library(cowplot); packageVersion("cowplot")
library(reshape); packageVersion("reshape")
library(write_xlsx); packageVersion("write_xlsx")
```

```{r housekeeping}
dir.create("plots")
dir.create("tables")
```

```{r read data}
Sys.time()
Sys.Date()

path_16s <- "16S/"
list.files(path_16s)
```

```{r plot 16s quality, fig.width=10, fig.height=5}
Sys.time()
Sys.Date()

# Forward and reverse fastq filenames have format: s01_1.fq and s01_2.fq
fnFs_16s <- sort(list.files(path_16s, pattern="_1", full.names = TRUE))
fnRs_16s <- sort(list.files(path_16s, pattern="_2", full.names = TRUE))
# Extract sample names, assuming filenames have format: s**_1_.fastq
sample.names_16s <- sapply(strsplit(basename(fnFs_16s), "_"), `[`, 1)
plotQualityProfile(fnFs_16s[1:35])
```

```{r filter 16s reads}
Sys.time()
Sys.Date()
# Place filtered files in filtered/ subdirectory
dir.create("16S/filtered")
filt_path_16s <- file.path(path_16s, "filtered") 
filtFs_16s <- file.path(filt_path_16s, paste0(sample.names_16s, "_r1_filt.fastq"))
filtRs_16s <- file.path(filt_path_16s, paste0(sample.names_16s, "_r2_filt.fastq"))

length(fnFs_16s)
length(fnRs_16s)

out_16s <- filterAndTrim(fnFs_16s, filtFs_16s, fnRs_16s, filtRs_16s, maxN=0, maxEE=c(2,5), truncQ=2, 
                     rm.phix=TRUE, compress=FALSE, multithread=FALSE) 
out_16s
errF_16s <- learnErrors(filtFs_16s, multithread=FALSE) # estimate error rates in forward sequences
errR_16s <- learnErrors(filtRs_16s, multithread=FALSE) # estimate error rates in reverse sequences

plotErrors(errF_16s, nominalQ=TRUE) # visualize estimated error rates
```

```{r 16s assign taxa}
Sys.time()
Sys.Date()
# dereplicate filtered fastq files
derepFs_16s <- derepFastq(filtFs_16s, verbose=TRUE)
derepRs_16s <- derepFastq(filtRs_16s, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs_16s) <- sample.names_16s
names(derepRs_16s) <- sample.names_16s

dadaFs_16s <- dada(derepFs_16s, err=errF_16s, multithread=FALSE) #Infer the sequence variants in each sample
dadaRs_16s <- dada(derepRs_16s, err=errR_16s, multithread=FALSE)
dadaFs_16s[[1]]

#merge denoised forward and reverse reads
mergers_16s <- mergePairs(dadaFs_16s, derepFs_16s, dadaRs_16s, derepRs_16s, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers_16s[[1]])

#construct sequence table
seqtab_16s <- makeSequenceTable(mergers_16s)
dim(seqtab_16s)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab_16s)))

#Remove chimeric sequences:
seqtab.nochim_16s <- removeBimeraDenovo(seqtab_16s, method="consensus", multithread=FALSE, verbose=TRUE)
dim(seqtab.nochim_16s)
sum(seqtab.nochim_16s)/sum(seqtab_16s)

#Track sequences through pipeline. See if there is one step that loses too many reads. 
getN_16s <- function(x) sum(getUniques(x))
track_16s <- cbind(out_16s, sapply(dadaFs_16s, getN_16s), sapply(mergers_16s, getN_16s), rowSums(seqtab_16s), rowSums(seqtab.nochim_16s))
colnames(track_16s) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track_16s) <- sample.names_16s
head(track_16s)

taxa_silva <- assignTaxonomy(seqtab.nochim_16s, "databases/silva_nr_v138_train_set.fa", multithread=FALSE)
taxa_silva <- addSpecies(taxa_silva, "databases/silva_species_assignment_v138.fa")

taxa_rdp <- assignTaxonomy(seqtab.nochim_16s, "databases/rdp_train_set_16.fa", multithread=FALSE)
taxa_rdp <- addSpecies(taxa_rdp, "databases/rdp_species_assignment_16.fa")

taxa_silva_print <- taxa_silva # Removing sequence rownames for display only
taxa_rdp_print <- taxa_rdp
rownames(taxa_silva_print) <- NULL
head(taxa_silva_print)
rownames(taxa_rdp_print) <- NULL
head(taxa_rdp_print)
```

```{r 16s plots basic}
Sys.time()
Sys.Date()

theme_set(theme_bw())

map <- read.table("map_file_MM_C.txt", header=TRUE, sep = "\t")
#mapping file
sample_data(map)
map <- sample_data(map)

# Assign rownames to be Sample ID's
rownames(map) <- map$SampleID
samples.out_16s <- rownames(seqtab.nochim_16s)

samdf_16s <- as.data.frame(map)

rownames(samdf_16s) <- samples.out_16s

ps_taxa_silva <- phyloseq(otu_table(seqtab.nochim_16s, taxa_are_rows=FALSE), 
               sample_data(samdf_16s), 
               tax_table(taxa_silva))
ps_taxa_silva

plot_richness(ps_taxa_silva, x="Dilution", measures=c("Shannon", "Simpson"), color="Dilution")

# Transform data to proportions as appropriate for Bray-Curtis distances
ps_prop_taxa_silva <- transform_sample_counts(ps_taxa_silva, function(otu) otu/sum(otu))
# to prevent error due to zero values in negative controls:
ps_prop_taxa_silva <- prune_samples(sample_sums(ps_prop_taxa_silva) >= 0, ps_prop_taxa_silva) 
nmds_bray_silva <- ordinate(ps_prop_taxa_silva, method="NMDS", distance="bray")

plot_ordination(ps_prop_taxa_silva, nmds_bray_silva, color="Dilution", title="Bray NMDS")

top20_silva <- names(sort(taxa_sums(ps_taxa_silva), decreasing=TRUE))[1:20]
ps_top20_silva <- transform_sample_counts(ps_taxa_silva, function(OTU) OTU/sum(OTU))
ps_top20_silva <- prune_taxa(top20_silva, ps_top20_silva)
plot_bar(ps_top20_silva, x="SampleID", fill="Genus") + facet_wrap(~Dilution, scales="free_x")

ps_silva <- transform_sample_counts(ps_taxa_silva, function(OTU) OTU/sum(OTU))

test <- ordinate(ps_silva, method="NMDS", distance="bray")
plot_ordination(ps_silva, test, color = "Dilution")
plot_bar(ps_taxa_silva, x="Dilution", fill="Genus") +# facet_wrap(~Dilution, scales="free_x") +
  theme(legend.position = 'none')

test1 <- prune_samples(sample_data(ps_taxa_silva)$Dilution == "D0", x = ps_taxa_silva)

plot_bar(test1, x="SampleID") +# facet_wrap(~Dilution, scales="free_x") +
  theme(legend.position = 'none')
sample_sums(ps_taxa_silva)

ps_melt <- psmelt(ps_taxa_silva)

ggplot(ps_melt, aes(x = SampleID, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity") +
  theme(legend.position = 'none') +
  facet_grid(. ~ Dilution)

```

```{r make otu table}
ps <- ps_prop_taxa_silva

taxa_names(ps)
n_seqs <- seq(ntaxa(ps))
len_n_seqs <- nchar(max(n_seqs))
taxa_names(ps) <- paste("Seq", formatC(n_seqs, 
                                       width = len_n_seqs, 
                                       flag = "0"), sep = "_")
taxa_names(ps)

wholetax <- do.call(paste, c(as.data.frame(tax_table(ps))
                             [c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")], 
                             sep = "__"))  # to distinguish from "_" within tax ranks

# turn the otu_table into a data.frame
otu_export <- as.data.frame(otu_table(ps))
tmp <- names(otu_export)

# paste wholetax and OTU_ids together
for(i in 1:length(tmp)){
  names(tmp)[i] = paste(wholetax[i], tmp[i], sep = "__")
}

# overwrite old names
names(otu_export) <- names(tmp)

head(otu_export)[5]

write_xlsx(otu_export,"OTU.xlsx")
```


```{r 16s alpha div}
## Work in progress

## sample, material, replicate, treatment, sample_type

ps_16s <- phyloseq(otu_table(seqtab.nochim_16s, taxa_are_rows=FALSE), 
               sample_data(samdf_16s), 
               tax_table(taxa_silva))

est_rich_16s <- estimate_richness(ps_16s,  measures=c("Shannon", "Simpson"))
est_rich_16s$merge <- rownames(est_rich_16s)
samdf_16s$merge <- rownames(samdf_16s)
est_rich_16s <- merge(est_rich_16s, samdf_16s, by = "merge")
est_rich_16s <- melt(est_rich_16s)

ggplot(est_rich_16s, aes(x = merge, y = value, color = value)) + 
  geom_point(size = 2) +
  # scale_shape_manual(values=c(1, 2)) +
  xlab("") + ylab("Alpha Diversity Measure") +
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank()) +
  theme(legend.title=element_blank()) +
  # scale_color_manual(values=c("#E69F00", "#56B4E9")) +
  theme(axis.text.x = element_text(size  = 9, angle = 90, hjust = 1, vjust = 0.5)) +
  facet_wrap(. ~ variable, scales="free_y")
# ggsave("plots/alpha_div_16s", dpi = 250)
```




Citations:
Quast C, Pruesse E, Yilmaz P, Gerken J, Schweer T, Yarza P, Peplies J, Glöckner FO (2013) The SILVA ribosomal RNA gene database project: improved data processing and web-based tools. Opens external link in new windowNucl. Acids Res. 41 (D1): D590-D596.

Yilmaz P, Parfrey LW, Yarza P, Gerken J, Pruesse E, Quast C, Schweer T, Peplies J, Ludwig W, Glöckner FO (2014) The SILVA and "All-species Living Tree Project (LTP)" taxonomic frameworks. Opens external link in new windowNucl. Acids Res. 42:D643-D648

Abarenkov, Kessy; Zirk, Allan; Piirmann, Timo; Pöhönen, Raivo; Ivanov, Filipp; Nilsson, R. Henrik; Kõljalg, Urmas (2020): UNITE general FASTA release for Fungi. Version 04.02.2020. UNITE Community. https://doi.org/10.15156/BIO/786368

Benjamin Callahan. (2017). RDP taxonomic training data formatted for DADA2 (RDP trainset 16/release 11.5) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.801828